.changeset/modern-horses-poke.md

```diff
@@ -0,0 +1,24 @@
+---
+'@backstage/plugin-catalog-backend-module-msgraph': minor
+---
+
+Improved performance by reducing the number of calls being made to the Microsoft Graph API.
+
+- Increased the number of rows returned at a time from 100 to 999.
+- Load user data in batches when getting users by group membership, rather than making individual calls per user.
+
+Improved logging to provide more information as the task progresses.
+
+Added support for loading only users and only groups
+If you don't add any configuration for groups or users, then they won't be ingested.
```

Comment:
Sorry, I should probably have pulled a bit more context from the linked issue into this PR template.  The big problem we have with MS graph is that for any medium to large org, the process takes many hours to run, without any sign of progress  (No logs, nothing showing up in the catalogue). This is a massive barrier to entry for users getting started with MS graph which involves a lot of support questions on github & discord, (There are probably even more users who just give up)

This PR contains a two pronged attack on this by 
1. Improve performance.  (If the ingestion is quick enough, the lack of progress issue becomes a non issue
2. Make it clearer that progress is happening, and make it easier to see what is going on.

From my experience with support questions, it is rarely the case that people want all users and especially not all groups ingested.  We get a lot of support questions on discord about the MS Graph ingestion not working, to which the answer is nearly always "Give it more time".

For me, I found the current behaviour surprising.  Take the following config.  My expectation would be that this config loads users by the filter I gave, but nothing else.  However hte current behaviour sees the missing group configuration section as an instruction to load all groups.
```yaml
catalog:
  providers:
    microsoftGraphOrg:
      default:
        tenantId: ${AZURE_TENANT_ID}
        userGroupMember:
          filter: id eq 'BLAH'
```

I could maybe get behind a setup where if no restrictiosn are specitifed at all, load all users and groups, but I still quite strongly think the behaviour of hte previous example loading all groups is not intuative. 

```yaml
catalog:
  providers:
    microsoftGraphOrg:
      default:
        tenantId: ${AZURE_TENANT_ID}
```

That said, I'm still not convinced that loading all users and groups is ever a good idea.

1. At best, I think most people want to ingest all *employees*.  Whilst you may logically think "employees.  Whilst all employees are users in AD, so too are other things.  (e.g. service accounts, external partners).  At my org, we have 10-20x more users objects than we do employees.  Even after filtering out disabled & guest accounts, we're still 5-10x more user objects than employees.  I don't have stats on other orgs, but from past experiences, I don't believe this is abnormal.
2. I also think it's pretty common to only want to ingest a specific subset of employees (e.g. tech teams)into backstage, and not the rest of the org

Restricting the default filter to just enabled & non guest accounts is a saner default than everyone, but still results in a lot of surprising users being ingested.  

There are also data privacy implications to loading unneeded users as users contain PII data (in the default transformer, names).  The GDPR principal of data minimisation says to not store any data you don't need.  It is unlikely your company's sales & marketing teams will use backstage, so you probably don't need to have their names stored in backstage.  The default of loading everyone will cause you to violate that principal by default.

Groups also suffer some surprising behaviour
1. My org has more groups than we do users.  (i.e. 10-20x the number of employees).  
2. Group names in AAD are not unique, but backstage uses this as their identifier.   As there are many different types of groups in AD, it's not uncommon to have groups with different names (e.g. there might be an SRE Team in MS Teams, an SRE email list (sre@company.org), an SRE security group (for permissions) and an SRE group for the team's SharePoint site.  Which one of these will backstage ingest as `group:sre`?  In an ideal world, each of these groups would have identical membership in AD, so it ultimately wouldn't matter as all 4 would ingest identically, however in reality the membership of the AD group usually drifts over time.   In most cases, you can avoid most of these conflicts by filtering to just one specific type of group, I don't think we can do that as a default as different orgs will vary by which type of group is best kept up to date.

I'm not sure of any sane default we can apply here.


If the MS graph plugin was fast enough, I might not be so worried about loading extra data.  However for both users & groups, there are some n+1 type queries required which mean that large numbers of extra users & groups have significant impact on the overall ingestion time.  (In some cases by several hours).
* For users, avatars have to be loaded by separate calls per user, rather than in the paginated lists
* For groups, backstage needs to load the group membership to fill int eh members field, resulting in more n+1 type queries.